### ISSUE TITLE
OpenAPI 3.1 support?

### ISSUE THREAD

**Issue by mitar** [deepmap/oapi-codegen#373]
OpenAPI 3.1 is great, brings full JSON Schema compatibility. It would be great if generator would support it.

**Comment by eli-kaplan** [comment]
Bumping this, OAS 3.1 [was released](https://www.openapis.org/blog/2021/02/18/openapi-specification-3-1-released) approximately seven months ago now and brings a number of very useful features.

**Comment by sonu27** [comment]
This would be great

**Comment by jamietanna** [comment]
So right now you can use a document with an OpenAPI 3.1 version, but we don't yet support new stuff.

There's a great post about what we need to add at https://www.openapis.org/blog/2022/05/11/openapi-v3-1-resources-for-tooling-developers

Is there anything we want to count as out of scope? I'd be tempted to look at whether I can help with the effort to add support

**Comment by jamietanna** [comment]
Related upstream issue that we'd need to wait for before we can support everything https://github.com/getkin/kin-openapi/issues/230

**Comment by jamietanna** [comment]
Per https://github.com/getkin/kin-openapi/issues/230#issuecomment-1284319206 this could also be a good option, but would lead to significant rework - may be worth it, especially as kin-openapi's not yet gotten around to it

**Comment by amery** [comment]
If you end up redesigning oapi-codegen using libopenapi, could you please consider allowing generating code for other languages too? I'm sure people here would be interested in using the same awesome tool to generate non-go client code

**Comment by jamietanna** [comment]
Just to cover this, there's some discussion between @deepmap-marcinr and I about this, and we're considering - for now - to convert an OpenAPI 3.1 spec to an OpenAPI 3.0 spec (via libopenapi) to continue to rely on kin-openapi, which will allow for OpenAPI 3.1 specs to continue to work, but without some of the benefits.

We'd prefer to keep on kin-openapi for the meantime, but also if there's no expected update to 3.1 (and folks do really want 3.1) we can look at shifting to use libopenapi, but it'll require a slightly more significant set of changes than we're willing to do right now, especially as [Marcin is continuing to think about what a v2 could look like](https://github.com/deepmap/oapi-codegen/discussions/440)

**Comment by jamietanna** [comment]
For folks looking for OpenAPI 3.1 support, what features are you hoping for? We are looking at what initial changes may be required (supporting multiple `type`s in a schema), but then wondering what additional things could be added to the backlog, like webhook support

**Comment by mgabeler-lee-6rs** [comment]
> For folks looking for OpenAPI 3.1 support, what features are you hoping for?

For my team's use cases, you nailed the two key items on the head -- `type` as an array, and `webhooks`, in that order of importance.

The type array could definitely be translated to an OAS 3.0 spec pretty easily, by converting it to a `oneOf` or `anyOf`, though nullability has some minor challenges in some cases.

**Comment by wjkoh** [comment]
The fully implemented version of OpenAPI 3.1 can be found at https://github.com/pb33f/libopenapi, while kin-openapi is still in the process of implementing it. Would switching to the former accelerate the support for OAS 3.1?

**Comment by jamietanna** [comment]
So using libopenapi is proposed in https://github.com/deepmap/oapi-codegen/pull/1388 and is a good option to support OpenAPI 3.1.

It's not something I want us to switch over to as-is, as it's a significant breaking change for folks (for those using this as a library, and some behaviour changes may occur) so would assume this would be a side-by-side thing for some time before maybe switching over.

Additionally, we also need to make sure the middleware used by the project is also https://github.com/pb33f/libopenapi-validator/ instead of using kin-openapi (as they'll differ in what's supported)

I'd welcome folks trying out #1388 with their repos, making sure that it works correctly with your services, etc, and we'll hopefully get it merged before long, but I want to make sure that we set ourselves up for it well due to i.e. needing to make sure that updates to internal code for generating templates also applies to the libopenapi side of things.

It's very much something I want to get done, just also trying to balance it with several other Open Source projects I maintain (all in my free time) as well as trying to keep on top of the incoming issues and PRs!

**Comment by guodongq** [comment]
any update for supporting openapi `3.1.0` ?

**Comment by jamietanna** [comment]
> any update for supporting openapi `3.1.0` ?

Unfortunately not - we've got a lot of "BAU" kinda issues/PRs to look into before we can look at strategic things that will allow us to push forwards towards supporting i.e. https://github.com/oapi-codegen/oapi-codegen/pull/1388 or an alternative approach.

Sponsorship is _hugely appreciated_ if this is blocking your organisations - and it's likely the only way we'll be able to put the time in towards it (related: https://github.com/oapi-codegen/oapi-codegen/discussions/1606)

**Comment by filiprejmus** [comment]
sponsoring a small monthly amount to support this

> > any update for supporting openapi `3.1.0` ?
> 
> Unfortunately not - we've got a lot of "BAU" kinda issues/PRs to look into before we can look at strategic things that will allow us to push forwards towards supporting i.e. #1388 or an alternative approach.
> 
> Sponsorship is _hugely appreciated_ if this is blocking your organisations - and it's likely the only way we'll be able to put the time in towards it (related: #1606)

**Comment by jamietanna** [comment]
> sponsoring a small monthly amount to support this

Thanks Filip! We appreciate that, and you putting your money where your mouth is!

Note that the sponsorship you'd sent through is under the 1hr work/mo tier, so may not directly lead to much time we can give towards this issue, but it should help alongside other contributions / in the long run

**Comment by willbuckner** [comment]
I know it's not much but I also put in a small sponsership for this one :) Appreciate the work you guys are doing and I hope it helps a bit.

**Comment by wikipeteia** [comment]
Bumping this, OAS 3.1.1 [was released this October](https://spec.openapis.org/oas/v3.1.1.html). Are there currently plans to support this?

**Comment by ajaypatel24** [comment]
Hey! Is there any way to follow or contribute to this work?

**Comment by jamietanna** [comment]
The update from https://github.com/oapi-codegen/oapi-codegen/issues/373#issuecomment-2424798110 still holds true, unfortunately.

To get OpenAPI 3.1 support in `oapi-codegen` is a good chunk of work, in planning and making sure we're set up in a way that the maintenance burden doesn't overload us, on top of our already high support levels - it's great we're heavily used, but also involves a tonne of work! 

https://github.com/oapi-codegen/oapi-codegen/pull/1388 is an option, but supporting both / migrating over is a significant chunk of work and I can't forsee being able to work on this without a significant time investment

**Comment by StevenACoffman** [comment]
If anyone else would like to contribute to this effort, consider getting your organization to sponsor the work, or you can personally [sponsor oapi-codegen](https://github.com/sponsors/oapi-codegen) or [jamietanna](https://github.com/sponsors/jamietanna).

The other way to contribute to this is to take PR #1388, fork it, update it, test it out, and contribute any fixes back as another PR.

**Comment by StevenACoffman** [comment]
Another approach would be to either contribute code or [sponsorship of fenollp](https://github.com/sponsors/fenollp) for the existing upstream library [kin-openapi](https://github.com/getkin/kin-openapi) to support OpenAPI 3.1 - see here https://github.com/getkin/kin-openapi/issues/230  and specifically [this comment](https://github.com/getkin/kin-openapi/issues/230#issuecomment-1002659224)

> I have been maintaining this project for a few years now and am looking to turn this into a more sustainable initiative.  Recently, I set up GitHub Sponsorship so if you'd like to support this project's future and preserve some of my mental health I encourage you to register for a monthly sponsoring [here](https://github.com/sponsors/fenollp). A one-time donation can also help solve piecemeal issues.
> 
> Here's a tutorial from GitHub on [how to find other maintainers that may need your help](https://docs.github.com/en/sponsors/getting-started-with-github-sponsors/quickstart-for-finding-contributors-to-sponsor) whose work you might depend on.
> 
> Regarding this specific issue, considering the large ramifications and attention to quality it entails, I would like to work on it once some funding has been promised or secured. We can discuss terms privately.
> 
> Licensing won't change due to sponsorship. All open-source work will continue to be available to all and enjoyed equally by all community members.

**Comment by jamietanna** [comment]
Moving [this comment](https://github.com/oapi-codegen/oapi-codegen/pull/1388#issuecomment-2492908737) here:

> Any plans to make a v3 tag with this breaking change?

This would be great - a good balance between moving folks to `libopenapi` while making it clear it's a breaking change.

When we performed the bump to v2 @mromaszewicz and I  decided we would not support v1, especially as it was not a significantly breaking change, and in our opinion would be a fairly low effort move (as we had seen on codebases we made the changes as tests) 

However, for the v3 bump - whether that's `libopenapi` or [some of the breaking changes planned for v3](https://github.com/oapi-codegen/oapi-codegen/milestone/4) - we do not want to follow this model. 

Instead, we will support v2 and v3 together, for some time I.e. 12 months. This will allow folks to move over to v3 more leisurely, while getting (at least) bug fixes. 

But the problem with that, linking into #1606, is that it doubles our support burden.

We would need to now manage two sets of versions and respective maintenance. Given we're struggling to keep on top of a single version - with the sponsorship we're super appreciative of - it wouldn't be fair to us as maintainers or y'all as community to say we're going to maintain v2 and v3 together but then not set us all for success.

Trust me, I'd _love_ to have this support in, and I'd _love_ to say we supported OpenAPI 3.1 with all that entails - but not at the (already high) cost of maintenance.

If we're able to get significantly more sponsors and dedicate more time to bringing down the backlog of outstanding issues and getting us to a more sustainable pace, then we'd be more happy to reconsider and see if we could create a v3

Hope y'all understand 💜

**Comment by jamietanna** [comment]
In the meantime, in https://github.com/oapi-codegen/oapi-codegen/pull/1966  we've added some documentation on how to use an OpenAPI 3.1 spec with `oapi-codegen` (downgrading in a less-painful way using OpenAPI Overlay)

As a reminder, sponsorship welcome (monthly or one-off!) to help put things in place to work on this large piece of work!

**Comment by ltwmaster** [comment]
@jamietanna can you put into context how much sponsorship for something such as this feature is needed to entirely complete it?

**Comment by jamietanna** [comment]
Folks looking for OpenAPI 3.1 support - what are the key features of 3.1 you're looking to take advantage of?

**Comment by fkarg** [comment]
@jamietanna My backend is FastAPI, which for a while now generates 3.1 by default. Some of the features I've had troubles with in the past:
- declaring parts of an enum as deprecated or adding other metadata (i.e. descriptions)
- proper support for oneOf / anyOf / allOf
- type unions
- additional descriptions at various places


I believe there's been some more, but these come to mind immediately.

**Comment by dharmjit** [comment]
Hey @jamietanna, as is the case for me and probably for many others starting new projects, folks generally prefer to use the recent versions of the standards even if the previous versions supports their current needs. I just wanted to highlight this as another data point to consider when prioritizing support for OAS 3.1.

**Comment by jamietanna** [comment]
Yep, understand - largely trying to see what particular features of 3.1 would be best to look at prioritising as part of the work, based on feedback

Similar to the rest of `oapi-codegen` where it's better to have examples of how different features are used, so we can then verify they work as expected

**Comment by akrieg-ionos** [comment]
In our project we had to downgrade a huge spec from 3.1. to 3.0 because of missing 3.1. support. It turned out that downgrade was just converting 
```json
"anyOf": [
      {
          "type": "<some type>",               
      },
      {
          "type": "null"
      }
   ],
```
to 
```json
    "type": "<some type>",
    "nullable": true,
```
At least for us , supporting `anyOf` with `"type": "null"` would already be sufficient.

**Comment by mattoni** [comment]
We have several specs here: https://github.com/cycleplatform/api-spec that we and our customers use extensively all written in 3.1. Currently I'm downgrading using a manual script I wrote to achieve it but its not a good long term solution.

The biggest things for us are
- proper support for oneOf / anyOf / allOf
- type unions

which are tricky to get right. Given the scope of our spec, I would encourage you to use it for testing. There may be a couple things we need to align on with it but overall it works great for documentation and our typescript client.

**Comment by NullHypothesis** [comment]
> Folks looking for OpenAPI 3.1 support - what are the key features of 3.1 you're looking to take advantage of?

This one, taken from the [3.1 release notes](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0-rc0):
> When using the Reference Object, summary and description fields can now be overridden.

This would allow us to make the description of referenced objects context-dependent, like so:
```yaml
type: object
properties:
  id:
    $ref: ./ID.yaml
    description: |
      This description overrides whatever is in ID.yaml,
      making it possible to provide context-dependent descriptions.
```

**Comment by rngtng** [comment]
FYI speakeasy open sourced their parser with OpenAPI 3.1.x support. Havn't looked deep insight yet, but maybe switching to that could bring the project further https://www.speakeasy.com/blog/release-oss-openapi-library?

**Comment by jamietanna** [comment]
Hey folks, unfortunately the sponsorship that was intended to be used for this (and the upstream work by Pierre in kin-openapi) has fallen through which is significantly disappointing 😞 I'll keep y'all posted on updates

Re using other libraries - be it [`libopenapi`](https://github.com/pb33f/libopenapi) or [`github.com/speakeasy-api/openapi`](https://github.com/speakeasy-api/openapi), the work to migrate is significant, especially given the breadth of bugs, features and compatibility we have right now.

It's unfortunately too much work to take on right now - or if it were to be something we did, would be something that would be the sole focus for a significant portion of the year and all other bugs/etc would be ignored

Not saying "no", but clarifying that there's a bit more to it, unfortunately

See also: https://github.com/oapi-codegen/oapi-codegen/issues/373#issuecomment-2493224049

See also: https://github.com/oapi-codegen/oapi-codegen/pull/1388 / https://github.com/oapi-codegen/oapi-codegen/pull/1211

**Comment by Jiehong** [comment]
and looks there is now openapi 3.2: https://learn.openapis.org/upgrading/v3.1-to-v3.2.html

It's mostly backward compatible with 3.1, so this ticket also represents this IMO

**Comment by GeertJohan** [comment]
There is a PR over at kin-openapi 😄   https://github.com/getkin/kin-openapi/pull/1102

**Comment by Daniel63656** [comment]
Does this mean this repo can be updated as well?

**Comment by jamietanna** [comment]
Yes, once Pierre (kin-openapi maintainer) and I have confirmed we've got sufficient coverage of OpenAPI 3.1 support, then `oapi-codegen` will be able to support OpenAPI 3.1 too

**Comment by zelch** [comment]
@jamietanna Is there anything that we can do to help with this?

If there's a branch of oapi-codegen that uses the updated kin-openapi, we could definitely test against our APIs.

**Comment by freetonik** [comment]
One way around this: https://www.jvt.me/posts/2025/05/04/oapi-codegen-trick-openapi-3-1/

**Comment by cubahno** [comment]
This repo was forked by Doordash and now open-sourced: https://github.com/doordash-oss/oapi-codegen-dd
`libopenapi` used as provider.
You are welcome to give it a try!

**Comment by dahu33** [comment]
> This repo was forked by Doordash and now open-sourced: [doordash-oss/oapi-codegen-dd](https://github.com/doordash-oss/oapi-codegen-dd) `libopenapi` used as provider. You are welcome to give it a try!

@cubahno Great work and thanks for sharing. Any plans to support server stub generation, even if only via user-provided templates?

**Comment by cubahno** [comment]
hey, ty! 
yes, server stubs will be there. probably within the next 3 months or so.

**Comment by mromaszewicz** [comment]
This is a very old issue, but I'm working on it.

I now have libopenapi based prototype (https://github.com/mromaszewicz/oapi-codegen/tree/experimental-v3/experimental) which generates models quite a bit better than the current kin-openapi version. I've avoided many mistakes from the past, but it's still a super early work in progress.

**Comment by mromaszewicz** [comment]
Please see the `experimental/` directory in this repo, and try out the new cmd/oapi-codegen in that directory.

It's feature complete and has many tests, but we need to hammer on this experiment for a while to know if it's good enough.

**Comment by x-delfino** [comment]
the experiment with libopenapi was moved to a fork btw incase anyones looking for it https://github.com/oapi-codegen/oapi-codegen-exp

**Comment by hmmh-dominik-saeume** [comment]
It's like 5 years now since 3.1 released. It is kind of crazy that it's still not supported in the official generator toolings.
That kind of sets a bad precedence discouraging adopting OAS.

So I guess if moonwalk ever releases, we will have to wait like another 10 years after that?
Or is there already some preemptive refactoring in process to ensure, we wont have the same happen with it?

**Comment by w0rp** [comment]
@hmmh-dominik-saeume As a FOSS author and president of a charity myself, I would insist that if you want a feature implemented, you should implement it. The forks above are the right approach. FOSS is always an act of charity.

**Comment by hmmh-dominik-saeume** [comment]
> [@hmmh-dominik-saeume](https://github.com/hmmh-dominik-saeume) As a FOSS author and president of a charity myself, I would insist that if you want a feature implemented, you should implement it. The forks above are the right approach. FOSS is always an act of charity.

So I'm correct to assume, that the foundation of OAS suggest using forks over the official tools?

Adoption of specs only work, if the tools match the spec.
It's literary the reason most of the specs still found online are 3.0 Instead of 3.1 or 3.2

When you're going on the OAS docs, you're reading 3.2 as it's the latest stable, and it links to official tooling that doesn't support the same version. This creates confusion and blocks adoption.

Isn't the goal to bring forward a standard with **matching** standard tooling?
How are projects supposed to confidently adopt OAS, if its missing core features, and adoption speed of their own specification isn't months but years?

**Comment by w0rp** [comment]
> So I'm correct to assume, that the foundation of OAS suggest using forks over the official tools?

For free software, that's exactly correct. Proprietary software is controlled by a software vendor and you cannot modify the code. If the vendor stops supporting it, you can't fix it. Free software allows you to modify your own copy to benefit you. Standardization lies in the format such as the OpenAPI Specification. The software tools are just code and binaries.

**Comment by hmmh-dominik-saeume** [comment]
The question was specifically about a foundation, not just a single project.

A Foundation is itself after all a body above the project itself.
I completely agree on normal FOSS, but as this lies under `oapi-codegen/` and is part of the foundation and is linked to everywhere in the docs, you would assume it matches the docs.
With the current state it seems like it can't even be seen as a reference implementation of the Specification any more, as it doesn't match.

**Comment by x-delfino** [comment]
the fork is because this is a pretty significant change and so they want to get some good testing done first - this makes sense imo. The experimental change is switching to libopenapi - that library has been pretty good at keeping up with previous releases:

| Spec Version  | Release Date | libopenapi Support Date |
| ------------- | ------------ | ----------------------- |
| OpenAPI 3.2.0 | 2025-09-19   | 2025-09-22              |
| OpenAPI 3.1.1 | 2024-10-24   | 2024-11-18              |

Hopefully making this change will mean that this project can keep up to date more easily. But generally I think it's pretty difficult when there are underfunded projects with people propping it up in their free time. I think the most direct way to help is contributing changes or sponsoring the project

**Comment by w0rp** [comment]
@x-delfino Yeah. Create the fork, test it, then submit a pull request.

**Comment by jamietanna** [comment]
Also, note that I'd explicitly moved `oapi-codegen`'s experimental support into a fork, as it should not have been in our main codebase to start with

We've still not decided if we are going to move away from kin-openapi - my personal preference is that we work with Pierre to make sure that kin-openapi is updated, rather than migrate away from it

(because there are other things in the ecosystem that are using kin-openapi and moving everyone on together is the best choice)

For **companies** that are wanting this, please also get in touch and sponsor the work for Pierre and use to work on this

Although Marcin's experimental fork is in progress, as I say, it's **not decided** on this being the approach going forwards

**Comment by w0rp** [comment]
> For companies that are wanting this, please also get in touch and sponsor the work for Pierre and use to work on this

Is there an appropriate place I can send USD to sponsor this work? I can talk to the Dense Analysis board about it.

**Comment by jamietanna** [comment]
For sponsorship, @fenollp and @oapi-codegen - Pierre, shall I reach out privately to you to confirm if we can share the amount of time (and therefore money) we quoted for the work? That may help give folks an indication of how much we're looking to need to "raise"

**Comment by vtopc** [comment]
> As a FOSS author and president of a charity

@w0rp, by the way. How did you manage to do so many breaking changes in the scope of one major version, between OpenAPI 3.0 and OpenAPI 3.1? All those broken `nullable`, multiple types, etc.

Sorry if off-topic.

**Comment by hmmh-dominik-saeume** [comment]
> > As a FOSS author and president of a charity
> 
> [@w0rp](https://github.com/w0rp), by the way. How did you manage to do so many breaking changes in the scope of one major version, between OpenAPI 3.0 and OpenAPI 3.1? All those broken `nullable`, multiple types, etc.
> 
> Sorry if off-topic.

isnt it supposed to be semantic versioning?
isnt it just a minor?

**Comment by jamietanna** [comment]
Please move this conversation to a separate discussion - this is more in scope of the OpenAPI working group

**Comment by GeertJohan** [comment]
For those interested, there is activity in https://github.com/getkin/kin-openapi/issues/230 and a new PR was made and merged https://github.com/getkin/kin-openapi/pull/1125 🎉

**Comment by mromaszewicz** [comment]
Oh, this is excellent.

I have a functional OpenAPI 3.1 oapi-codegen in `../oapi-codegen-exp`, and I'll be able to use AI to backport the 3.1 idioms from that experiment to this repo.

**Comment by mromaszewicz** [comment]
OpenAPI 3.1 support is working in the branch `feat/kin-openapi-3.1`. All backend routers are supported, but we can't merge this for a while (see the commits and description in the above PR for more info)

This code will not pass CI until we use a Go 1.26 linter and limit CI to go 1.26, as this version of kin-openapi requires go 1.26

However, for those interested, feel free to experiment.

**Comment by mromaszewicz** [comment]
kin-openapi has backed off their Go requirement, and now we need to bump to 1.25. That's already out for review, and `feat/kin-openapi-3.1` has been updated as well.

**Comment by Stanzilla** [comment]
> kin-openapi has backed off their Go requirement, and now we need to bump to 1.25. That's already out for review, and `feat/kin-openapi-3.1` has been updated as well.

I just tried your branch it fails on on nullable object schema `(type: [object,'null'])`

**Comment by mromaszewicz** [comment]
> > kin-openapi has backed off their Go requirement, and now we need to bump to 1.25. That's already out for review, and `feat/kin-openapi-3.1` has been updated as well.
> 
> I just tried your branch it fails on on nullable object schema `(type: [object,'null'])`

Thanks for pointing that out. I had unit tests covering various cases, but I missed `[object, 'null']` for unspecified objects. The issue stems from the `Schema.Is(...)` function returning "false" for an object. It's a strict comparison. I'll have it fixed shortly, with additional tests. I found five places in the code which could have problems due to this behavior which I misread.

**Comment by mromaszewicz** [comment]
@Stanzilla , please try the latest commit.

Also, henceforth, since people are checking this branch out, I will merge `main` into here, versus rebasing on top of `main`.

**Comment by cristiadu** [comment]
Any news on when this would be able to be released?

**Comment by mromaszewicz** [comment]
I'm going to make another pass at new bugs in the next few days, and decide then. This branch has some problems around nullable types. All the places that I compute a type has to be aware of both the 3.1 and 3.0 description, and I'm not sure if I covered all the bases yet.
